presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -16,53 +16,284 @@
+    {
+        // TODO convert to a pattern that allows for a sequence of ProjectNode, instead
+        // of a canned number, once the pattern system supports it.
+        return IntStream.range(0, 5)
+                .boxed()
+                .flatMap(numProjects ->
+                        Stream.of(
+                                new MergeAdjacentWindowsOverProjects(numProjects),
+                                new SwapAdjacentWindowsBySpecifications(numProjects)))
+                .collect(toImmutableSet());
+    }
+
+    private abstract static class ManipulateAdjacentWindowsOverProjects
+            implements Rule<WindowNode>
+    {
+        private final Capture<WindowNode> childCapture = newCapture();
+        private final List<Capture<ProjectNode>> projectCaptures;
+        private final Pattern<WindowNode> pattern;
 
-    private static final Pattern<WindowNode> PATTERN = window()
-            .with(source().matching(window().capturedAs(CHILD)));
+        protected ManipulateAdjacentWindowsOverProjects(int numProjects)
+        {
+            PropertyPattern<PlanNode, ?> childPattern = source().matching(window().capturedAs(childCapture));
+            ImmutableList.Builder<Capture<ProjectNode>> projectCapturesBuilder = ImmutableList.builder();
+            for (int i = 0; i < numProjects; ++i) {
+                Capture<ProjectNode> projectCapture = newCapture();
+                projectCapturesBuilder.add(projectCapture);
+                childPattern = source().matching(project().capturedAs(projectCapture).with(childPattern));
+            }
+            this.projectCaptures = projectCapturesBuilder.build();
+            this.pattern = window().with(childPattern);
+        }
+
+        @Override
+        public Pattern<WindowNode> getPattern()
+        {
+            return pattern;
+        }
 
-    @Override
-    public Pattern<WindowNode> getPattern()
+        @Override
+        public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
+        {
+            // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+            // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+            // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+            // input symbols.
+            List<ProjectNode> projects = projectCaptures.stream()
+                    .map(captures::get)
+                    .collect(toImmutableList());
+
+            return pullUnaryNodeAboveProjects(
+                    WindowNode.class,
+                    captures.get(childCapture),
+                    projects)
+                    .flatMap(newChild -> manipulateAdjacentWindowNodes(parent, newChild, context));
+        }
+
+        public abstract Optional<PlanNode> manipulateAdjacentWindowNodes(WindowNode parent, WindowNode child, Context context);
```

Comment:
private?

Reply:
I'll make it protected, which I think is what you mean.

Reply:
hm... I thought about `private`. As both belong to the same file.